### PR TITLE
Support for NoExecute taint, pod affinity, and bug fix volume attachment delete

### PR DESCRIPTION
**Pull request by rbo54**
# Description
This PR contains:
1. Changes to the podmontest helm script to support pod affinity and pod topology spread in order to test customer use case.
2. A new script bounce.kubelet to test scenarios where a kubelet is not present.
3. Changes to the controller code to process pods with affinity together when discovered by ArrayConnectivityLoss to try and help the k8s scheduler make better decisions.
4. A change to the controller logic to only honor the existing flag SkipArrayConnectionValidation when the pod's node is tainted with a k8s No Execute taint (meaning kubernetes wants to evacuate the node.) This makes the code safer.
5. A retry that fixes the problem in CSM-148 unable to delete volume attachment.
6. A new flag on nway.sh "rebalance" that will rebalance the pod to node distribution after a failover if required rather than failing because some nodes need more than 110 pods. Also a separate rebalance.sh script to do the same thing.
# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| [CSM-148](https://github.com/dell/csm/issues/148) - Fix for unable to delete volume attachment  |
 [CSM-87](https://github.com/dell/csm/issues/87) - enhancement so that enabling SkipArrayConnectionValidation is safer as is only done when the node has a No Execute taint|
| [CSM-165](https://github.com/dell/csm/issues/165) - CSM Resiliency enhancement to consider pod affinity  |

# Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [x] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Backward compatibility is not broken

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- [x] New unit tests for pod affinity
- [x] Executed VxFlex full integration test- results:
- 18 scenarios (18 passed)
184 steps (184 passed)
1h33m23.697942654s
time="2022-01-20T20:18:43-05:00" level=info msg="Integration test finished"
--- PASS: TestPowerFlexIntegration (5603.73s)
PASS

### CODE REVIEW

**Comment by rbo54** [comment]
I will fix the forbidden words with a grep of 'mast.r' instead of what kubernetes calls them.

File: cmd/podmon/main.go [hunk 1]
```diff
@@ -160,7 +160,7 @@ func main() {
 				go ArrayConnMonitorFc()
 			}
 			// monitor all the nodes with no label required
-			go StartNodeMonitorFn(K8sAPI, k8sapi.K8sClient.Client, "", "", monitor.MonitorRestartTimeDelay)
+			// XXX go StartNodeMonitorFn(K8sAPI, k8sapi.K8sClient.Client, "", "", monitor.MonitorRestartTimeDelay)
```
Comment (tdawe) [inline cmd/podmon/main.go:163]: Should this not be commented out?

Comment (rbo54) [reply cmd/podmon/main.go]: Yes, will fix.

**Review by alikdell — APPROVED** [review]

**Review by shaynafinocchiaro — DISMISSED** [review]

**Review by tdawe — DISMISSED** [review]

**Review by sharmilarama — APPROVED** [review]

File: internal/monitor/controller.go [hunk 2]
```diff
@@ -403,6 +454,23 @@ func (cm *PodMonitorType) ArrayConnectivityMonitor() {
 	}
 }
 
+// ProcessPodInfoForCleanup processes a ControllerPodInfo for cleanup, checking that the UID and object are the same, and then calling controllerCleanupPod.
+func (cm *PodMonitorType) ProcessPodInfoForCleanup(podInfo *ControllerPodInfo, reason string) {
```
Comment (alikdell) [inline internal/monitor/controller.go:458]: This func is not required to expose!

**Comment by rbo54** [comment]
I have updated the PR, addressing Trevor's concern. I have run full integration test for VxFlex and Unity, both passed. Unit test still also passing.

**Review by alikdell — DISMISSED** [review]

**Review by shaynafinocchiaro — APPROVED** [review]

**Review by tdawe — APPROVED** [review]

**Review by alikdell — APPROVED** [review]

**Review by shaynafinocchiaro — APPROVED** [review]

**Pull request merged**